### PR TITLE
Set runspec on all new tasks to avoid deadlocks

### PR DESCRIPTION
**Pull request by fjetter**
This is a more minimalistic version of #4360 with fixes I am pretty confident with. I still can see local issue, though and this is incomplete.

### CODE REVIEW

**Comment by mrocklin** [comment]
I notice that this is still in WIP status.  Are the changes here an improvement over what is in master?  If so, I might press the "Ready for review" button

**Comment by gforsyth** [comment]
I'll note that the failures in `test_failed_workers.py::test_worker_who_has_clears_after_failed_connection` are connected to work stealing -- if stealing is disabled the assertion errors go away

**Comment by fjetter** [comment]
I'm currently testing the latest commit on a few clusters in parallel. So far, no funny logs, no deadlocks, no problems at all.

**Comment by fjetter** [comment]
The only thing I could observe, so far is that the waiting_for_data_count is off but I checked the workers, both data and tasks are empty and I suspected a bug in this counter already. I believe this is not critical, though, since the counter is not used internally, merely for instrumentation

image

**Comment by mrocklin** [comment]
That's great news

**Comment by fjetter** [comment]
Needed to push another time. There was actually a typo on my side causing a test to fail.

**Comment by fjetter** [comment]
Unfortunately, I needed to push another small change but an expected one where we didn't clear `who_has` on the task state during `release_key`. That's not a big deal, I believe, and my tests so far showed this works still nicely.

I'm dropping out now but if there is no big CI thing popping up again, I feel confident merging this now. I'm not sure if it fixes all issues but it should provide a significant improvement to master from what I've seen.

**Comment by mrocklin** [comment]
@fjetter are there other issues that you know exist?  If so I'm happy to take a look at them today.

File: distributed/worker.py [hunk 1]
```diff
@@ -1691,9 +1699,26 @@ def transition_ready_executing(self, ts):
...
         self.send_task_state_to_scheduler(ts)
 
+    def transition_ready_waiting(self, ts):
+        """
+        This transition is common for work stealing
+        """
+        ts.state = "waiting"
+        ts.runspec = None
+
+        return ts.state
```
Comment (mrocklin) [inline distributed/worker.py:1720]: Do we need to transition to waiting in this case?  I'm curious if this was necessary before due to the transition when we were releasing keys, but maybe it's not necessary now?

Comment (mrocklin) [reply distributed/worker.py:1720]: Also, in both cases where we call this the final state is "waiting".  Do we need to set the state explicitly, or can we let the `transition` function handle this?

**Comment by fjetter** [comment]
On the other PR I originally wanted to fix an issue with erroneous tasks which led me to discover this runspec thing. I added a test case with an object with broken deserialization. That should hit the worker internal suspicious counter (but it doesn't/didn't) something else might reveal itself there. Otherwise I don't have anything on my mind right now

File: distributed/worker.py (hunk 1 above)
Comment (fjetter) [reply distributed/worker.py]: This transition is probably not necessary. The only case I am aware of where we hit this is work stealing.

**Comment by mrocklin** [comment]
OK, once tests pass here I'm going to merge.  If things are ok for a bit I suggest that we release.  cc @jrbourbeau

**Comment by mrocklin** [comment]
I briefly tried things with the exercise in #4430 and things didn't complete. I need to double check that work though.

**Comment by fjetter** [comment]
I had already a few successful runs of the example in #4430 and that worked out great. That was using an earlier commit of this PR and I'll double check with master now.

Really happy to see this merged. Thanks for your help @mrocklin !

**Comment by mrocklin** [comment]
Ah, indeed, things with that workflow are good for me too as well now.  Good for a release whenever makes sense @jrbourbeau .  I suspect that you're aiming for this Friday.

**Comment by jrbourbeau** [comment]
xref https://github.com/dask/community/issues/121

**Pull request merged**